shiny/workflowPlot/bm.R

```diff
@@ -0,0 +1,275 @@
+        ))
+      )
+    }
+  })
+  if(bm$ready){bm$calc_bm_message <- sprintf("Please select at least one variable and one metric")}
+})
+
+observeEvent({
+  input$vars
+  input$metrics
+  input$plots
+  },{
+  v <- ifelse(is.null(input$vars),0,length(input$vars))
+  n <- ifelse(is.null(input$metrics),0,length(input$metrics))
+  p <- ifelse(is.null(input$plots),0,length(input$plots))
+  m <- n + p
+  output$report <- renderText(sprintf("Number of vars: %0.f, Number of metrics: %0.f", v,m))
+  if(v > 0 & m > 0){
+    output$calc_bm_button <- renderUI({actionButton("calc_bm", "Calculate Benchmarks")})
+    bm$bm_vars <- input$vars
+    bm$bm_metrics <- c()
+    if(n > 0) bm$bm_metrics <- c(bm$bm_metrics, input$metrics)
+    if(p > 0) bm$bm_metrics <- c(bm$bm_metrics, input$plots)
+  }
+  
+}, ignoreNULL = FALSE)
+
+observeEvent(input$calc_bm,{
+  req(input$all_input_id)
+  req(input$all_site_id)
+  req(bm)
+  bm$calc_bm_message <- sprintf("Setting up benchmarks")
+  output$reportvars <- renderText(paste(bm$bm_vars, seq_along(bm$bm_vars)))
+  output$reportmetrics <- renderText(paste(bm$bm_metrics))
+  
+  inputs_df <- getInputs(bety,c(input$all_site_id)) %>% 
+    dplyr::filter(input_selection_list == input$all_input_id)
+  output$inputs_df_table <- renderTable(inputs_df)
+  
+  config.list <- PEcAn.utils::read_web_config("../../web/config.php")
+  output$config_list_table <- renderTable(as.data.frame.list(config.list))
+  
+  bm$bm_settings$info <- list(userid = 1000000003) # This is my user id. I have no idea how to get people to log in to their accounts through the web interface and right now the benchmarking code has sections dependent on user id - I will fix this. 
+  bm$bm_settings$database <- list(
+    bety = list(
+      user = config.list$db_bety_username,
+      password = config.list$db_bety_password,
+      host = config.list$db_bety_hostname,
+      dbname = config.list$db_bety_database,
+      driver = config.list$db_bety_type,
+      write = TRUE
+    ),
+    dbfiles = config.list$dbfiles_folder
+  )
+  bm$bm_settings$benchmarking <- list(
+    ensemble_id = bm$ens_wf$ensemble_id,
+    new_run = FALSE
+  )
+  
+  for(i in seq_along(bm$bm_vars)){
+    benchmark <- list(
+      input_id = inputs_df$input_id,
+      variable_id = bm$bm_vars[i],
+      site_id = inputs_df$site_id,
+      metrics = list()
+    )
+    for(j in seq_along(bm$bm_metrics)){
+      benchmark$metrics = append(benchmark$metrics, list(metric_id = bm$bm_metrics[j]))
+    }
+    bm$bm_settings$benchmarking <- append(bm$bm_settings$benchmarking,list(benchmark = benchmark))
+  }
+  
+  output$calc_bm_button <- renderUI({})
+  output$print_bm_settings <- renderPrint(bm$bm_settings)
+  
+  basePath <- dplyr::tbl(bety, 'workflows') %>% dplyr::filter(id %in% bm$ens_wf$workflow_id) %>% dplyr::pull(folder)
+  
+  settings_path <- file.path(basePath, "pecan.BENCH.xml")
+  saveXML(PEcAn.utils::listToXml(bm$bm_settings,"pecan"), file = settings_path)
```

Comment:
This may be useful for debugging, but ultimately I don't think you need to write this out. Can't you just call define_benchmark(bm$bm_settings) right here?

Reply:
Yes - right now I'm keeping it for debugging but it's not necessary.